pkg/dns/target.go

```diff
@@ -153,6 +153,70 @@ func (t *ClusterGatewayTarget) setGeo(defaultGeo GeoCode) {
 	t.Geo = &geoCode
 }
 
+func (t *MultiClusterGatewayTarget) RemoveUnhealthyGatewayAddresses(probes []*v1alpha1.DNSHealthCheckProbe, listener gatewayv1beta1.Listener) {
+
+	//If we have no probes we can't determine health so return unmodified
+	if len(probes) == 0 {
+		return
+	}
+
+	//Build a map of gateway addresses and their health status
+	gwAddressHealth := map[string]bool{}
+	allunhealthy := true
+	for _, cgt := range t.ClusterGatewayTargets {
+		for _, gwa := range cgt.GatewayAddresses {
+			gwProbes := getProbeForGatewayAddress(probes, gwa, t.Gateway.Name, string(listener.Name))
+			//if all probes are unhealthy return as we want to include all endpoints
+			if len(gwProbes) > 1 {
+				fmt.Println("more than 1 probe found")
+			}
+			for _, probe := range gwProbes {
+				probeHealthy := true
+				if probe.Status.Healthy != nil {
+					probeHealthy = *probe.Status.Healthy
+				}
+				if probeHealthy && probe.Spec.FailureThreshold != nil && probe.Status.ConsecutiveFailures < *probe.Spec.FailureThreshold {
+					allunhealthy = false
+				}
+				gwAddressHealth[gwa.Value] = probeHealthy
+			}
+		}
+	}
+	//If we have no matching probes for our current addresses, or we have no healthy probes, return unmodified
+	if len(gwAddressHealth) == 0 || allunhealthy {
+		return
+	}
+
+	// Remove all unhealthy addresses, we know by this point at least one of our addresses is healthy
+	for _, cgt := range t.ClusterGatewayTargets {
+		healthyAddresses := []gatewayv1beta1.GatewayAddress{}
+		for _, gwa := range cgt.GatewayAddresses {
+			if healthy, exists := gwAddressHealth[gwa.Value]; exists && healthy {
+				healthyAddresses = append(healthyAddresses, gwa)
+			}
+		}
+		cgt.GatewayAddresses = healthyAddresses
+	}
+}
+
+func getProbeForGatewayAddress(probes []*v1alpha1.DNSHealthCheckProbe, gwa gatewayv1beta1.GatewayAddress, gatewayName, listenerName string) []*v1alpha1.DNSHealthCheckProbe {
+	retProbes := []*v1alpha1.DNSHealthCheckProbe{}
+	for _, probe := range probes {
+		if dnsHealthCheckProbeName(gwa.Value, gatewayName, listenerName) == probe.Name {
+			retProbes = append(retProbes, probe)
+		}
+	}
+	return retProbes
+}
+
+func dnsHealthCheckProbeName(address, gatewayName, listenerName string) string {
+	return fmt.Sprintf("%s-%s", address, dnsRecordName(gatewayName, listenerName))
+}
+
+func dnsRecordName(gatewayName, listenerName string) string {
+	return fmt.Sprintf("%s-%s", gatewayName, listenerName)
+}
```

Comment:
Need to find a better way to manage these as they are currently duplicated as cannot import the package.

Reply:
Hmm, not ideal. We really shouldn't be using the name as the unique id anyway, labels and selectors would probably be a better way of selecting matching resources, not that that helps here.

You could at least get rid of the duplication of `dnsRecordName` by changing `dnsHealthCheckProbeName` to `fmt.Sprintf("%s-%s", address, gatewayName, listenerName)`, your not gaining much by calling it.

Reply:
the current setup of labels and selectors doesn't provide what we need to pick up the correct probename. We could refactor that. Would you preference be to do that here or given we are using this method currently on main should we create another JIRA for it? i don't mind either way.

Reply:
Yeah i'd say just leave it using the name and we look at it later. DNSRecord has the same issue, we should really be doing the same thing for all created resources.

Reply:
Captured this here https://github.com/Kuadrant/multicluster-gateway-controller/issues/534